### ISSUE TITLE
no gradients for torch.tensor(np.array(...)) on cuda

### ISSUE THREAD

**Issue by maremun** [pytorch/pytorch#7335]
Hello, below is the code that yields no gradients for the tensor with `requires_grad=True`. Doesn't work with `dtype=torch.double` neither. Also, when using anything from numpy for `data` argument, e.g. `torch.tensor.(np.ones_like(something), ...)`.

```
a = torch.tensor(np.array([1.,2.,3.]), dtype=torch.float, requires_grad=True).cuda()
print(a.requires_grad, a.dtype)  # True torch.float32
b = torch.ones(3, device='cuda')
l = a@b
print(l.item())  # 6.0
l.backward()
print(a.grad)  # None
```
PyTorch version 0.4.0
Python 3.6.5

**Comment by zou3519** [comment]
You're hiding the reference to the original tensor you created:
```
a = torch.tensor(np.array([1.,2.,3.]), dtype=torch.float, requires_grad=True).cuda()
```
This creates a tensor that gradients accumulates into, and then you call `.cuda()` on it, which creates a second tensor. 

Try 
```
a = torch.tensor(np.array([1.,2.,3.]), dtype=torch.float, device="cuda", requires_grad=True)
```
instead.
